<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

# Project 1: Python Coding Exercises

_Authors: Joseph Nelson (DC) _

---

The following code challenges are drawn from common exercises used in technical interviews.

Please note that there may be several ways to approach each challenge. If you get stuck, try mapping out your approach in pseudocode first. Finally, while solutions to problems like these may be found online, remember that if you copy/paste code that you can't explain, you'll be missing out on the point of the project. The only way to truly learn a new skill is through practice, trial, and error - we can only help you improve by understanding where you are having trouble.

### Challenge 1: Largest Palindrome
A palindromic number reads the same both ways. For example, 1234321 is a palindrome. The largest palindrome made from the product of two two-digit numbers is 9009 = 91 × 99. Find the largest palindrome made from the product of two three-digit numbers. Afterward, write a brief explanation walking through your code's logic in markdown.

In [1]:
import numpy as np
range(1,1000)

range(1, 1000)

In [2]:
p=[]
for i in range(1,1000):
    for j in range(1,1000):
        p1=i*j
        p2 = str(p1)
        p3 = p2[::-1]
        if p2==p3:
            p.append(p1)
            
max(p)

906609

p = palindrome

The largest palindrome made from product of two three-digit numbers is 906609 = 913 x 993. Using two for loop to get the range of the numbers because to find out which three-digit numbers will get the maximum numbers. Change it into string so that it can be reverse. To know which number is the maximum, used max() so it can execute the maximum number and we will get the largest palindrome numbers made from product of two three-digit numbers.


### Challenge 2: Summation of Primes
The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17. Find the sum of all the primes below 2,000. Afterward, write a brief explanation walking through your code's logic in markdown.

In [5]:
# to find all prime numbers below 2000
start = int (input("Enter the starting range:"))
end = int (input("Enter the end range:"))
print ("Prime numbers in the range", start, "to", end)
for i in range(start, end+1):
    flag = 0
    for j in range(2, i):
        if (i % j == 0):
            flag = 1
            break
            
    if (flag == 0):
        print (i, end = ' ')

Enter the starting range:2
Enter the end range:2001
Prime numbers in the range 2 to 2001
2 3 5 7 11 13 17 19 23 29 31 37 41 43 47 53 59 61 67 71 73 79 83 89 97 101 103 107 109 113 127 131 137 139 149 151 157 163 167 173 179 181 191 193 197 199 211 223 227 229 233 239 241 251 257 263 269 271 277 281 283 293 307 311 313 317 331 337 347 349 353 359 367 373 379 383 389 397 401 409 419 421 431 433 439 443 449 457 461 463 467 479 487 491 499 503 509 521 523 541 547 557 563 569 571 577 587 593 599 601 607 613 617 619 631 641 643 647 653 659 661 673 677 683 691 701 709 719 727 733 739 743 751 757 761 769 773 787 797 809 811 821 823 827 829 839 853 857 859 863 877 881 883 887 907 911 919 929 937 941 947 953 967 971 977 983 991 997 1009 1013 1019 1021 1031 1033 1039 1049 1051 1061 1063 1069 1087 1091 1093 1097 1103 1109 1117 1123 1129 1151 1153 1163 1171 1181 1187 1193 1201 1213 1217 1223 1229 1231 1237 1249 1259 1277 1279 1283 1289 1291 1297 1301 1303 1307 1319 1321 1327 1361 1367 1373 1381 139

In [7]:
#to calculate the sum of all the prime numbers below 2000

max=int(input("Enter the maximum value for find sum of primes:"))
sum=0
for num in range(2,max+1):
    i=2
    for i in range(2,num):
        if(int(num%i==0)):
            i=num
            break;

    if i is not num:
        sum+=num
        
print("Sum of all prime numbers 2 to ",max,":",sum)

Enter the maximum value for find sum of primes:2000
Sum of all prime numbers 2 to  2000 : 277048


The first method is to find out all prime numbers below 2000. Next, in order to get the sum of all prime numbers below 2000, need to enter the maximum value of the prime and the range start from 2 since 2 is the minimum number of prime. It will print out the sum of all prime numbers from 2 to 2000.

### Challenge 3: Multiples of 3 and 5
If we list all of the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6, and 9. The sum of these multiples is 23. Find the sum of all the multiples of 3 and 5 below 1,000. Afterward, write a brief explanation walking through your code's logic in markdown.

In [8]:
# to list all the numbers below 1000 that are multiples of 3 or 5
multiples = []
a = 1
while a < 1000:
    if a % 3 == 0:
        multiples.append(a)
    elif a % 5 == 0:
        multiples.append(a)
    a = a + 1
    
print(multiples)

[3, 5, 6, 9, 10, 12, 15, 18, 20, 21, 24, 25, 27, 30, 33, 35, 36, 39, 40, 42, 45, 48, 50, 51, 54, 55, 57, 60, 63, 65, 66, 69, 70, 72, 75, 78, 80, 81, 84, 85, 87, 90, 93, 95, 96, 99, 100, 102, 105, 108, 110, 111, 114, 115, 117, 120, 123, 125, 126, 129, 130, 132, 135, 138, 140, 141, 144, 145, 147, 150, 153, 155, 156, 159, 160, 162, 165, 168, 170, 171, 174, 175, 177, 180, 183, 185, 186, 189, 190, 192, 195, 198, 200, 201, 204, 205, 207, 210, 213, 215, 216, 219, 220, 222, 225, 228, 230, 231, 234, 235, 237, 240, 243, 245, 246, 249, 250, 252, 255, 258, 260, 261, 264, 265, 267, 270, 273, 275, 276, 279, 280, 282, 285, 288, 290, 291, 294, 295, 297, 300, 303, 305, 306, 309, 310, 312, 315, 318, 320, 321, 324, 325, 327, 330, 333, 335, 336, 339, 340, 342, 345, 348, 350, 351, 354, 355, 357, 360, 363, 365, 366, 369, 370, 372, 375, 378, 380, 381, 384, 385, 387, 390, 393, 395, 396, 399, 400, 402, 405, 408, 410, 411, 414, 415, 417, 420, 423, 425, 426, 429, 430, 432, 435, 438, 440, 441, 444, 445, 447, 450,

In [9]:
sum = 0
for i in multiples:
    sum = sum + i
print(sum)

233168


Firstly, I try to list all the numbers that are multiples of 3 or 5 and start to sum up all the numbers after I get the output.

### Challenge 4: String Compressor
Implement a method to perform basic string compression using the counts of repeated characters. (This is called run-length encoding.) For example, the string "aabcccccaaa" would become a2b1c5a3. If the “compressed” string would not become smaller than the original string, your method should return the original string. You can assume the string has only uppercase and lowercase letters (a–z). Specify whether your solution is case sensitive or case insensitive and what you would need to change to make it the other. Afterward, write a brief explanation walking through your code's logic in markdown.

In [10]:
def string(message):
    string_message = ""
    i = 0
   
    while (i <= len(message)-1):
        cnt = 1
        char = message[i]
        j = i
        while (j < len(message)-1):
            if (message[j] == message[j+1]):
                cnt = cnt+1
                j = j+1
            else:
                break
                
        string_message=string_message+char+str(cnt)
        i = j+1
        
    return string_message

In [12]:
string_message=string("aabcccccaaa")
print(string_message)

a2b1c5a3


I used 'string' to define the message to encode. Use while if because if the next character is similar to the current character, the count will be increase but if it is not similar, the while loop will be break. In order to print the character and its count, I include the char and str along with the message.

### *BONUS* Challenge: FizzBuzz
Write a program that prints all of the numbers from 1 to 100. For multiples of 3, instead of the number, print "Fizz;" for multiples of 5, print "Buzz." For numbers that are multiples of both 3 and 5, print "FizzBuzz." Afterward, write a brief explanation walking through your code's logic in markdown.

In [13]:
for num in range(1, 101):
    if num % 3 == 0 and num % 5 == 0:
        print("fizzbuzz")
    elif num % 3 == 0:
        print("fizz")
    elif num % 5 == 0:
        print("buzz")
    else:
        print(num)

1
2
fizz
4
buzz
fizz
7
8
fizz
buzz
11
fizz
13
14
fizzbuzz
16
17
fizz
19
buzz
fizz
22
23
fizz
buzz
26
fizz
28
29
fizzbuzz
31
32
fizz
34
buzz
fizz
37
38
fizz
buzz
41
fizz
43
44
fizzbuzz
46
47
fizz
49
buzz
fizz
52
53
fizz
buzz
56
fizz
58
59
fizzbuzz
61
62
fizz
64
buzz
fizz
67
68
fizz
buzz
71
fizz
73
74
fizzbuzz
76
77
fizz
79
buzz
fizz
82
83
fizz
buzz
86
fizz
88
89
fizzbuzz
91
92
fizz
94
buzz
fizz
97
98
fizz
buzz


All number that were list are divisible by 3 and 5 but were change into 'fizz' and 'buzz'. If the number are multiples of 3 and 5, 'fizzbuzz' will be printed. 